<a href="https://colab.research.google.com/github/ElenaSaveleva/Project_2_SQL/blob/main/Project_2_%D0%9D%D0%BE%D1%83%D1%82%D0%B1%D1%83%D0%BA_%D0%A1%D0%B0%D0%B2%D0%B5%D0%BB%D1%8C%D0%B5%D0%B2%D0%B0.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [111]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [112]:
import pandas as pd
import psycopg2
import plotly.express as px

In [114]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 


In [115]:
connection = psycopg2.connect(
   dbname=DBNAME,
   user=USER,
   host=HOST,
   password=PASSWORD,
   port=PORT
)

In [116]:
query_1 = f'''select *  
                FROM public.vacancies
            '''    
hh_df = pd.read_sql_query(query_1, connection)
display(hh_df.head()) 
display(hh_df.info()) # информация о числе непустых значений в столбцах и их типах (vacancies)


id                                               name  \
0  55312386                                Компьютерный Мастер   
1  55843014                            Системный администратор   
2  54525964                      Lead Java Developer to Poland   
3  54525965                      Lead Java Developer to Poland   
4  55354053  Специалист службы поддержки с техническими зна...   

                                          key_skills          schedule  \
0  Пользователь ПК\tРабота в команде\tРемонт ноут...       Полный день   
1  Средства криптографической защиты информации\t...       Полный день   
2    Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
3    Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
4                                               None  Удаленная работа   

      experience           employment  salary_from  salary_to  area_id  \
0      Нет опыта     Полная занятость      64000.0        NaN     1516   
1  От 3 до 6 лет     Полная занятость          NaN        NaN      130   
2  От 3 до 6 лет     Полная занятость          NaN        NaN      160   
3  От 3 до 6 лет     Полная занятость          NaN        NaN      159   
4      Нет опыта  Частичная занятость      15000.0        NaN     1955   

   employer_id  
0      5724811  
1      4903713  
2        69961  
3        69961  
4         1740

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49197 entries, 0 to 49196
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           49197 non-null  int64  
 1   name         49197 non-null  object 
 2   key_skills   40741 non-null  object 
 3   schedule     49197 non-null  object 
 4   experience   49197 non-null  object 
 5   employment   49197 non-null  object 
 6   salary_from  21485 non-null  float64
 7   salary_to    12690 non-null  float64
 8   area_id      49197 non-null  int64  
 9   employer_id  49197 non-null  int64  
dtypes: float64(2), int64(3), object(5)
memory usage: 3.8+ MB


None

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [117]:
# текст запроса
query_1 = f'''select count (id) as all_vacancies
                from public.vacancies
            ''' 

In [118]:
# результат запроса
hh_df = pd.read_sql_query(query_1, connection)
print(f'Количество вакансий в базе (table vacancies): {hh_df.iloc[0,0]}')

Количество вакансий в базе (table vacancies): 49197


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [119]:
# текст запроса
query_2 = f'''select count (id) as all_employers
                from public.employers 
            '''    

In [120]:
# результат запроса
hh_df2 = pd.read_sql_query(query_2, connection)
print(f'Количество работадателей в базе (table employers): {hh_df2.iloc[0,0]}') 

Количество работадателей в базе (table employers): 23501


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [121]:
# текст запроса
query_3 = f'''select count (id) as all_areas
                from public.areas 
            '''    

In [122]:
# результат запроса
hh_df3 = pd.read_sql_query(query_3, connection)
print(f'Количество регионов в базе (table areas): {hh_df3.iloc[0,0]}') 

Количество регионов в базе (table areas): 1362


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [123]:
# текст запроса
query_4 = f'''select count (id) as all_industries
                from public.industries 
            '''    

In [124]:
# результат запроса
hh_df4 = pd.read_sql_query(query_4, connection)
print(f'Количество сфер деятельности в базе (table industries): {hh_df4.iloc[0,0]}') 

Количество сфер деятельности в базе (table industries): 294


***

In [125]:
# выводы по предварительному анализу данных

#### В данной базе данных представлено достаточно много ваканесий (49197), что позволяет соискателю подобрать подходящую для него вакансию. 
#### Вакансии представлены в 1392 регионах, что дает возможность соискателю найти работу по месиу проживания. 
####Работадателей в базе представлено  23501, а следовательно широкий спектр требований и условий при подборе вакансий для соискателя. 
####В базе 294 сферы деятельности. Разнообразие сфер деятельности тоже дает большие возможности при выборе походящего места работы. 

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [126]:
# текст запроса
query_4_1 = f'''select 
                      a.name as area, 
                      count(v.id) as cnt_vacancies
                  from public.areas a
                  join public.vacancies v on v.area_id = a.id
                  group by area
                  order by cnt_vacancies desc
'''

In [127]:
# результат запроса
hh_df_4_1 = pd.read_sql_query(query_4_1, connection)
display(hh_df_4_1)

area  cnt_vacancies
0             Москва           5333
1    Санкт-Петербург           2851
2              Минск           2112
3        Новосибирск           2006
4             Алматы           1892
..               ...            ...
764       Тарко-Сале              1
765    Новоаннинский              1
766            Бирск              1
767           Сасово              1
768             Анна              1

[769 rows x 2 columns]

Для большей наглядности полученных данных о количестве вакансий в каждом регионе посторим столбчатую диаграмму

In [128]:
# Посторим график
data = hh_df_4_1[hh_df_4_1['cnt_vacancies']>100]
fig = px.bar(data,
    x = "area",         
    y = "cnt_vacancies",
    width = 800,
    height = 500,
    title = "Распределение вакансий по регионам (мин.количество вакансий 100)",
    barmode = "group",
    orientation='v',
    color = 'cnt_vacancies'
)
fig.show()
fig.write_html('/content/drive/MyDrive/data/plotly/vacancies.html')

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [129]:
# текст запроса
query_4_2 = f'''select count(id) as cnt_vacancies
              from public.vacancies 
              where salary_from IS NOT NULL or salary_to IS NOT NULL
'''

In [130]:
# результат запроса
hh_df_4_2 = pd.read_sql_query(query_4_2, connection)
display(hh_df_4_2)

cnt_vacancies
0          24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [131]:
# текст запроса
query_4_3 = f'''select
                      round(avg(salary_from)) as avg_salary_from,
                      round(avg(salary_to)) as avg_salary_to
              from public.vacancies
'''

In [132]:
# результат запроса
hh_df_4_3 = pd.read_sql_query(query_4_3, connection)
display(hh_df_4_3)

avg_salary_from  avg_salary_to
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [133]:
# текст запроса
query_4_4 = f'''select distinct
                              schedule, 
                              employment, 
                              count(id) as cnt_vacancies
                from public.vacancies 
                group by schedule, employment
                order by cnt_vacancies desc 
'''

In [134]:
# результат запроса
hh_df_4_4 = pd.read_sql_query(query_4_4, connection)
display(hh_df_4_4)

schedule           employment  cnt_vacancies
0        Полный день     Полная занятость          35367
1   Удаленная работа     Полная занятость           7802
2      Гибкий график     Полная занятость           1593
3   Удаленная работа  Частичная занятость           1312
4     Сменный график     Полная занятость            940
5        Полный день           Стажировка            569
6     Вахтовый метод     Полная занятость            367
7        Полный день  Частичная занятость            347
8      Гибкий график  Частичная занятость            312
9        Полный день     Проектная работа            141
10  Удаленная работа     Проектная работа            133
11     Гибкий график           Стажировка            116
12    Сменный график  Частичная занятость            101
13  Удаленная работа           Стажировка             64
14     Гибкий график     Проектная работа             18
15    Сменный график           Стажировка             12
16    Вахтовый метод     Проектная работа              2
17    Сменный график     Проектная работа              1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [135]:
# текст запроса
query_4_5 = f'''select experience, count(id) as cnt_vacancies
                from public.vacancies 
                group by experience
                order by cnt_vacancies asc 
'''

In [136]:
# результат запроса
hh_df_4_5 = pd.read_sql_query(query_4_5, connection)
display(hh_df_4_5)

experience  cnt_vacancies
0         Более 6 лет           1337
1           Нет опыта           7197
2       От 3 до 6 лет          14511
3  От 1 года до 3 лет          26152

In [137]:
# выводы по детальному анализу вакансий

####Благодаря детельному анализу данных можно сделать следующие выводы;
####1. Наибольшее количество вакансий сконцентрировано в больших городах. Возможно это связано с высокой численостью населения и нехваткой кадров по сравнению с менее крупными городами. Лидерами здесь представлены такие города как, Москва, Санкт-Петербург, Минск, Новосибирск, Алматы.
####2. Зарплата является одним из важнейших условий для соискателя при выборе вакансии. В данной базе заполнено хотя бы одно из двух полей с зарплатой в 24073 вакансиях, что составляет почти половину от всех вакансий в базе (48,9%).
####3. Среднее значение для нижней и верхней границы зарплатной вилки равно : 71 065 руб. и 110 537 руб. соответственно, но для полноты анализа я бы сравнила с медианной, для этого была расчитана медиана представленных значений и получены следующие данные, которые существенно отличаются от среднего значения: 50 000 руб. и 80 000 руб.; данные показатели наиболее объективно отображают  распределение рассмотренных признаков, т.к. главное преимущество медианы в устойчивости к выбросам — аномально большим или малым значениям.
####4. При выявлении количества вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment) лидирующие позиции занимает пара "полный день - полная занятость" (72% от общего числа вакансий), далее вариант "удаленная работа - полная занятость" (16%), значительно ниже доля от общего числа вакансий у сочетания "гибкий график - полная занятость" (3%). Из этого следует, что большинство работадателей предлагают вакансии в качестве основного места работы. Также, в связи с современными тенденциями и обстоятельствами, появилась возможность для удаленной формы занятости на полный рабочий день. Данный вариант позволяет рассматривать такие вакансии соискателям из других регионов
####5. Благодаря тому, что работадатели предъявляют различные требования к опыту (от специалиста до новичка), соискательможет найти  вакансию в подходящей сфере. Большинство в данной базе принадлежит вакансиям с небольшим требуемым опытом работы - от 1 года до 3 лет (53.2% от общего числа вакансий). Как правило, данный сегмент ориентирован на соискателей, которые нацеленны на дальнейшее саморазвитие и карьерный рост.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [138]:
# текст запроса
query_5_1 = f'''select e.name employer, count(v.id) as cnt_vacancies
                from public.employers e
                join public.vacancies v on v.employer_id = e.id 
                group by employer
                order by cnt_vacancies desc
                limit 5
'''

In [139]:
# результат запроса
hh_df_5_1 = pd.read_sql_query(query_5_1, connection)
display(hh_df_5_1)

employer  cnt_vacancies
0         Яндекс           1933
1     Ростелеком            491
2       Тинькофф            444
3           СБЕР            428
4  Газпром нефть            331

Для большей наглядности полученных данных посторим столбчатую диаграмму

In [140]:
# Посторим график
fig = px.bar(hh_df_5_1,
    x = "employer",         
    y = "cnt_vacancies",
    width = 800,
    height = 500,
    title = "Лидеры по количеству вакансий",
    barmode = "group",
    orientation='v',
    color = 'cnt_vacancies'
)
fig.show()
fig.write_html('/content/drive/MyDrive/data/plotly/lider.html')

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [141]:
# текст запроса
query_5_2 = f'''select a.name as area, 
                       count(e.id) as cnt_employers,
                       count(v.id) as cnt_vacancies
                from public.areas a
                left join public.employers e on e.area = a.id
                left join public.vacancies v on v.area_id = a.id 
                where v.id IS NULL
                group by a.name
                order by cnt_vacancies desc, cnt_employers desc
'''

In [142]:
# результат запроса
hh_df_5_2 = pd.read_sql_query(query_5_2, connection)
display(hh_df_5_2)

area  cnt_employers  cnt_vacancies
0                            Россия            410              0
1                         Казахстан            207              0
2                Московская область             75              0
3                Краснодарский край             19              0
4                          Беларусь             18              0
..                              ...            ...            ...
588                        Кувандык              0              0
589                        Куйбышев              0              0
590  Куйбышево (Ростовская область)              0              0
591                        Култаево              0              0
592                        Кульсары              0              0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [143]:
# текст запроса
query_5_3 = f'''select e.name as employer, 
                       count(distinct (v.area_id)) as cnt_areas
                from public.employers e
                join public.vacancies v on v.employer_id = e.id 
                where v.id IS NOT NULL
                group by employer
                order by cnt_areas desc
'''

In [144]:
# результат запроса
hh_df_5_3 = pd.read_sql_query(query_5_3, connection)
display(hh_df_5_3)

employer  cnt_areas
0                      Яндекс        181
1                  Ростелеком        152
2                  Спецремонт        116
3      Поляков Денис Иванович         88
4                   ООО ЕФИН          71
...                       ...        ...
14761                  UniSol          1
14762            UNISTORY LLC          1
14763                   UNIT6          1
14764     United Distribution          1
14765              UNITEL LLC          1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [145]:
# текст запроса
query_5_4 = f'''select count(e.id) 
                from public.employers e
                left join public.employers_industries ei on e.id = ei.employer_id
                left join public.industries i on ei.industry_id = i.id
                where ei.employer_id IS NULL 
'''

In [146]:
# результат запроса
hh_df_5_4 = pd.read_sql_query(query_5_4, connection)
display(hh_df_5_4)

count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [147]:
# текст запроса
query_5_5 = f'''select e.name as employers_name
                from public.employers e
                left join public.employers_industries ei on e.id = ei.employer_id
                left join public.industries i on ei.industry_id = i.id
                group by employers_name
                having count(i.id) = 4
                order by employers_name
                offset 2 limit 1    
'''

In [148]:
# результат запроса
hh_df_5_5 = pd.read_sql_query(query_5_5, connection)
display(hh_df_5_5)

employers_name
0           2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [149]:
# текст запроса
query_5_6 = f'''select count(e.id)
                from public.employers e 
                left join public.employers_industries ei ON e.id = ei.employer_id -- соединяем необходимые таблицы через табл.employers_industries
                left join public.industries i ON ei.industry_id = i.id
                where i.name = 'Разработка программного обеспечения' 
              '''   


In [150]:
# результат запроса
hh_df_5_6 = pd.read_sql_query(query_5_6, connection)
display(hh_df_5_6)

count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [151]:
# Получим список городов-милионников
table_cities = pd.read_html('https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8')
df_city = table_cities[1]
cities = list(df_city["Город"]) # преобразуем необходимый столбец в список
cities[cities.index('Воронеж[a]')] = 'Воронеж' # вносим изменения в полученный список
cities_tuple = tuple(cities) # преобразуем в кортеж


In [152]:
# текст запроса
query_5_7 = f'''select 
                    a.name as cities_name, 
                    count(v.id) as cnt_vacancies
                from public.vacancies v 
                left join public.areas a on a.id = v.area_id 
                left join public.employers e on v.employer_id = e.id 
                where e.name = 'Яндекс' and a.name in {cities_tuple} --обращаемся к списку городов-миллионеров
                group by cities_name
               
                UNION ALL --добавим графу 'Total'
                
                select
                    'Total',
                    count(v.id) as cnt_vacancies
                from public.vacancies v 
                left join public.areas a on a.id = v.area_id 
                left join public.employers e on v.employer_id = e.id 
                where e.name = 'Яндекс' and a.name in {cities_tuple}
                order by cnt_vacancies
                '''   


In [153]:
# результат запроса
hh_df_5_7 = pd.read_sql_query(query_5_7, connection)
display(hh_df_5_7)

cities_name  cnt_vacancies
0              Омск             21
1         Челябинск             22
2        Красноярск             23
3         Волгоград             24
4             Пермь             25
5            Казань             25
6    Ростов-на-Дону             25
7               Уфа             26
8            Самара             26
9         Краснодар             30
10          Воронеж             32
11      Новосибирск             35
12  Нижний Новгород             36
13     Екатеринбург             39
14  Санкт-Петербург             42
15           Москва             54
16            Total            485

***

In [154]:
# выводы по анализу работодателей

####После детального анализа было выявлено 5 лидеров среди работадателей по количесиву вакансий. Это крупнейшие федеральные компании: Яндекс, Ростелеком, Тинькофф, СБЕР, Газпром нефть. На их долю приходится 7.4% от общего числа вакансий при общем количестве работодателей 23 501. Возможно это связано со значительным по численности персоналом компаний, большим количеством  территориальных филиалов этих компаний. 
#### К регионам с большим количеством работадателей, где отсутствуют вакансии, относятся те регионы, в которых спрос в данной сфере значительно превышает предложение со стороны работодателей. В связи с этим необходимо рассмотреть варианты трудоустройства для соискателей в других регионах с возможностью удаленной работы/переезда/командировок.
####Территориальное распределение вакансий у различных компаний очень неравномерно. Так, лидерующие позиции по числу регионов, в которых работодатель публикует свои вакансии, занимает Яндекс, на втором месте находится компания .Ростелеком. Это может быть связано с большим количеством вакансий в данных компаниях, наличием территориальных филиалов, а также может быть обусловлено нехваткой кадров в крупных городах.
####В имеющейся базе данных 8 419 работодателей (36% от общего числа) не указали свою сферу деятельности, что может снизить эффективность подбора вакансий, так как это является определяющим успешность поискового запроса вакансий для соискателей, 
####Также стоит отметить, что в имеющемся перечне вакансий есть и работодатели, которые указали несколько и более сфер деятельности, что, в свою очередь, значительно расширяет критерии поиска для соискателей, учитывая их индивидуальные особенности.
####3 553 (15% от общего числа) указали в качестве сферы деятельности "Разработка программного обеспечения". Данная сфера деятельности в большей степени ориентирована на IT-специалистов. 
#### Анализ имеющихся вакансий Яндекса, крупнейшего работодателя в нашей базе данных по общему количеству вакансий и по числу регионов с опубликованными вакансиями, в разрезе представленности в регионах-миллионниках показал, что на их долю приходится 25% от общего числа вакансий компании (485 позиций). Данный показатель подтверждает тенденцию в развитии компании в плане расширения региональной представленности кадров.


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [155]:
# текст запроса
query_6_1 = f'''select count(id) as cnt_vacancies
            from public.vacancies
            where lower(name) like '%data%' or lower(name) like '%данн%'
'''

In [156]:
# результат запроса
hh_df_6_1 = pd.read_sql_query(query_6_1, connection)
display(hh_df_6_1)

cnt_vacancies
0           1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [157]:
# текст запроса
query_6_2 = f'''select count(id)
                from public.vacancies
                where (lower(name) like '%data scientist%' or
                       lower(name) like '%data science%' or
                       lower(name) like '%исследователь данных%' or
                       (name like '%ML%' and name not like '%HTML%') or
                       lower(name) like '%machine learning%' or
                       lower(name) like '%машинн%обучен%') and
                       (lower(name) like '%junior%' or
                       experience = 'Нет опыта' or
                       employment = 'Стажировка')
'''

In [158]:
# результат запроса
hh_df_6_2 = pd.read_sql_query(query_6_2, connection)
display(hh_df_6_2)

count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [159]:
# текст запроса
query_6_3 = f'''select 
                      count(id) cnt_vacancies
                from public.vacancies
                where (lower(name) like '%data scientist%' or
                       lower(name) like '%data science%' or
                       lower(name) like '%исследователь данных%' or
                       (name like '%ML%' and name not like '%HTML%') or
                       lower(name) like '%machine learning%' or
                       lower(name) like '%машинн%обучен%') and
                       (key_skills like '%SQL%' or key_skills like '%postgres%')
'''

In [160]:
# результат запроса
hh_df_6_3 = pd.read_sql_query(query_6_3, connection)
display(hh_df_6_3)

cnt_vacancies
0            201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [161]:
# текст запроса
query_6_4 = f'''select 
                      count(id) cnt_vacancies
                from public.vacancies
                where (lower(name) like '%data scientist%' or
                       lower(name) like '%data science%' or
                       lower(name) like '%исследователь данных%' or
                       (name like '%ML%' and name not like '%HTML%') or
                       lower(name) like '%machine learning%' or
                       lower(name) like '%машинн%обучен%') and
                       (key_skills like '%Python%')
'''

In [162]:
# результат запроса
hh_df_6_4 = pd.read_sql_query(query_6_4, connection)
display(hh_df_6_4)

cnt_vacancies
0            351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [163]:
# текст запроса
query_6_5 = f'''select round(avg(length(key_skills) - length(replace(key_skills, CHR(9), ''))+1), 2) 
                       from public.vacancies
                       where lower(name) like '%data scientist%' or
                       lower(name) like '%data science%' or
                       lower(name) like '%исследователь данных%' or
                       (name like '%ML%' and name not like '%HTML%') or
                       lower(name) like '%machine learning%' or
                       lower(name) like '%машинн%обучен%' and
                       (key_skills is not null)
'''

In [164]:
# результат запроса
hh_df_6_5 = pd.read_sql_query(query_6_5, connection)
display(hh_df_6_5)

round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [165]:
# текст запроса
query_6_6 = f'''select
                    distinct experience,
                    round(avg(coalesce((salary_from+salary_to)/2, salary_from, salary_to)), 0) AS mean_salary
                from public.vacancies
                where (lower(name) like '%data scientist%' or
                       lower(name) like '%data science%' or
                       lower(name) like '%исследователь данных%' or
                       (name like '%ML%' and name not like '%HTML%') or
                       lower(name) like '%machine learning%' or
                       lower(name) like '%машинн%обучен%') and
                       (salary_from is not null or salary_to is not null) and
                       experience is not null
                group by experience       
'''

In [166]:
# результат запроса
hh_df_6_6 = pd.read_sql_query(query_6_6, connection)
display(hh_df_6_6)

experience  mean_salary
0           Нет опыта      74643.0
1  От 1 года до 3 лет     139675.0
2       От 3 до 6 лет     243115.0

***

In [167]:
# выводы по предметному анализу

####В данной базе непосредственное отношение к данным (вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн') имеют 1 771 вакансия (3,6% от общего количества вакансий). Из них только 480 вакансий (чуть менее 1% от общего количества вакансий), которые будут, в первую очередь, предлагаться соискателям, претендующим на позицию Data Scientist. В связи с этим можно сделать вывод, что имеющаяся база предлагает достаточно небольшой выбор вакансий, связанных с этой должностью.
####В вакансиях в качестве требований к претендентам работодатели указывали ключевые навыки, наличие которых необходимо для получение положительного решения при рассмотрении вакансии. В среднем в вакансиях по DS указано 6 навыков.
####Основными ключевыми навыками являются - владение Python (73% от общего числа вакансий на позицию DS), SQL или postgres (42% от общего числа вакансий на позицию DS).
####При расчете средней заработной платы специалиста DS в зависимости от имеющегося опыта работы были получены следующие результаты: для вакансий без опыта работы - 74 643 руб., от 1 года до 3 лет - 139 675 руб., от 3 до 6 лет - 243 115 руб. Мы наблюдаем большие значения для специалиста DS в связи с востребованностью направления в различных сферах, динамичным развитием данной области, нехваткой квалифцированных специалистов.


# Общий вывод по проекту

In [168]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

#### В данной базе данных содержится достаточно большое количество вакансий, которые размещены в различных регионах, с большим перечнем работодателей, имеющим различные требования к соискателям. В дальнейшем для размещения вакансий работодателем необходимо внести как можно больше полей с выбором параметров, значительно снизить долю заполняемых вручную.
####Графа name в таблице areas содержит одновременно и названия стран, и названия регионов, и отдельных городов, что мешает детально и верно проанализировать распределение вакансий по странам или регионам.
####Для более эффективного подбора вакансий также можно добавить в имеющуюся базу следующие требования к кандитатам: пол, возраст, образование.
####При выявлении наиболее популярных сочетаний типа рабочего графика и типа трудоустройства для вакансий DS в имеющихся вакансиях абсолютное большинство принадлежит варианту "полный день - полная занятость" (67% от общего числа вакансий DS), для варианта "удаленная работа - полная занятость" показатели чуть выше, чем для данных по всем вакансиям, что можно объяснить спецификой рассматриваемого направления (23 % от общего числа вакансий DS).
####При этом количество вакансий для начинающих специалистов в сфере DS также минимально (5,8% от общего числа вакансий DS), наиболее востребованы специалисты с опытом работы также от 1 года до 3 лет.


# Юнит 7. Дополнительные исследования данных

1. Какое количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment) для вакансий "Data Scientist" представлено в базе данных

In [169]:
# текст запроса
query_7_1 = f'''select distinct 
                    schedule, 
                    employment, 
                    count(id) as cnt_vacancies
                from public.vacancies  
                where (lower(name) like '%data scientist%' or
                       lower(name) like '%data science%' or
                       lower(name) like '%исследователь данных%' or
                       (name like '%ML%' and name not like '%HTML%') or
                       lower(name) like '%machine learning%' or
                       lower(name) like '%машинн%обучен%')
                group by schedule, employment
                order by cnt_vacancies DESC 
              '''

In [170]:
# результат запроса
hh_df_7_1 = pd.read_sql_query(query_7_1, connection)
display(hh_df_7_1)

schedule           employment  cnt_vacancies
0       Полный день     Полная занятость            319
1  Удаленная работа     Полная занятость            110
2     Гибкий график     Полная занятость             37
3     Гибкий график  Частичная занятость              4
4  Удаленная работа  Частичная занятость              3
5     Гибкий график           Стажировка              2
6       Полный день           Стажировка              2
7  Удаленная работа     Проектная работа              2
8       Полный день  Частичная занятость              1

2. В каких регионах представлены вакансии "Data Scientist"


In [171]:
# текст запроса
query_7_2 = f'''select
                     a.name as area, 
                     count(v.id) as cnt_vacancies
              from public.areas a 
              join  public.vacancies v on v.area_id = a.id
              where (lower(v.name) like '%data scientist%' or
                     lower(v.name) like '%data science%' or
                     lower(v.name) like '%исследователь данных%' or
                     (v.name like '%ML%' and v.name not like '%HTML%') or
                     lower(v.name) like '%machine learning%' or
                     lower(v.name) like '%машинн%обучен%') 
              group by area
              order by cnt_vacancies desc
'''   


In [172]:
# результат запроса
hh_df_7_2 = pd.read_sql_query(query_7_2, connection)
display(hh_df_7_2)

area  cnt_vacancies
0                      Москва            217
1             Санкт-Петербург             64
2                 Новосибирск             23
3             Нижний Новгород             20
4                      Казань             16
5                      Алматы             16
6                       Минск              9
7                       Томск              8
8                Екатеринбург              7
9                   Краснодар              6
10             Ростов-на-Дону              6
11                    Армения              6
12                 Нур-Султан              6
13                     Самара              5
14                      Пермь              5
15                       Кипр              5
16                     Рязань              4
17                    Ташкент              4
18                  Иннополис              3
19                    Тбилиси              3
20                Владивосток              3
21                     Сербия              3
22                       Омск              3
23                        Уфа              2
24                    Израиль              2
25                     Турция              2
26                      Пенза              2
27  Киров (Кировская область)              2
28                      Чехия              2
29                       Баку              2
30                      Литва              2
31                    Воронеж              2
32           Набережные Челны              1
33                    Барнаул              1
34                     Батуми              1
35                   Белгород              1
36                     Бишкек              1
37                     Брянск              1
38                  Волгоград              1
39                 Зеленоград              1
40                     Ижевск              1
41                    Королев              1
42                 Красноярск              1
43                    Абхазия              1
44                       Орел              1
45               Петрозаводск              1
46                     Польша              1
47                    Саратов              1
48                      Тверь              1
49                     Тюмень              1
50                  Челябинск              1
51                 Черногория              1

In [ ]:
connection.close()